test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala

```diff
@@ -0,0 +1,329 @@
+      }
+      for (v <- Seq(Long.MinValue, -1L, 2L, Long.MaxValue)) {
+        expectThrows(classOf[DateTimeException], t.`with`(ERA, v))
+      }
+    }
+  }
+
+  @Test def plus(): Unit = {
+    for {
+      t <- samples
+      u <- ChronoUnit.values() if isSupported(u)
+    } {
+      assertSame(t, t.plus(0, u))
+    }
+
+    assertEquals(Year.of(10), lastBCYear.plus(1, DECADES))
+    assertEquals(Year.of(100), lastBCYear.plus(1, CENTURIES))
+    assertEquals(Year.of(1000), lastBCYear.plus(1, MILLENNIA))
+    assertEquals(Year.of(1), lastBCYear.plus(1, ERAS))
+
+    assertEquals(Year.of(11), firstACYear.plus(1, DECADES))
+    assertEquals(Year.of(101), firstACYear.plus(1, CENTURIES))
+    assertEquals(Year.of(1001), firstACYear.plus(1, MILLENNIA))
+
+    assertEquals(lastBCYear, min.plus(999999999L, YEARS))
+    assertEquals(lastBCYear, max.plus(-999999999L, YEARS))
+
+    val values = Seq(Long.MinValue, Int.MinValue.toLong, -1000L, -366L, -365L,
+      -100L, -12L, -10L, -7L, -1L, 0L, 1L, 7L, 10L, 12L, 100L,
+      365L, 366L, 1000L, Int.MaxValue.toLong, Long.MaxValue)
+
+    for {
+      t <- samples
+      v <- values
+    } {
+      testDateTime(t.plus(v, YEARS))(t.plusYears(v))
+      testDateTime(t.plus(v, DECADES))(t.plusYears(Math.multiplyExact(v, 10)))
+      testDateTime(t.plus(v, CENTURIES))(t.plusYears(Math.multiplyExact(v, 100)))
+      testDateTime(t.plus(v, MILLENNIA))(t.plusYears(Math.multiplyExact(v, 1000)))
+    }
+
+    expectThrows(classOf[DateTimeException], firstACYear.plus(1, ERAS))
+    expectThrows(classOf[DateTimeException], lastBCYear.plus(-1, ERAS))
+  }
+
+  @Test def plusYears(): Unit = {
+    for (t <- samples) {
+      assertSame(t, t.plusYears(0))
+    }
+
+    assertEquals(firstACYear, lastBCYear.plusYears(1))
+    expectThrows(classOf[DateTimeException], min.plusYears(-1))
+    expectThrows(classOf[DateTimeException], max.plusYears(1))
+  }
+
+  @Test def minusYears(): Unit = {
+    for (t <- samples) {
+      assertSame(t, t.minusYears(0))
+    }
+
+    assertEquals(lastBCYear, firstACYear.minusYears(1))
+    expectThrows(classOf[DateTimeException], min.minusYears(Long.MinValue))
+    expectThrows(classOf[DateTimeException], max.minusYears(Long.MaxValue))
+  }
+
+  @Test def adjustInto(): Unit = {
+    val aDate = LocalDate.of(2015, 1, 1)
+    assertEquals(LocalDate.of(0, 1, 1), lastBCYear.adjustInto(aDate))
+  }
+
+  @Test def until(): Unit = {
+    for {
+      t <- samples
+      unit <- ChronoUnit.values() if isSupported(unit)
+    } {
+      assertEquals(0L, t.until(t, unit))
+    }
+
+    assertEquals(1999999998L, min.until(max, YEARS))
+    assertEquals(-1999999998L, max.until(min, YEARS))
+    assertEquals(1L, lastBCYear.until(firstACYear, YEARS))
+
+    val otherLeapYear: Year = {
+      var y = leapYear.getValue + 1
+      while (!Year.isLeap(y)) y += 1
+      Year.of(y)
+    }
+    assertEquals(4L, leapYear.until(otherLeapYear, YEARS))
```

Comment:
I would suggest using a tailrec function rather than a while.
```scala 
@tailrec def otherLeapYear(y: Int): Year = {
  val nextYear = y + 1
  if (Year.isLeap(nextYear)) Year.of(nextYear)
  else nextLeapYear(nextYear)
}
assertEquals(4L, leapYear.until(otherLeapYear(leapYear), YEARS))
```